ToolBox/PsychSignalDataConverter/PsychSignalDataDownloader.cs

```diff
@@ -0,0 +1,170 @@
+*/
+
+using System;
+using System.IO;
+using System.Net;
+using System.Threading;
+using NodaTime;
+using QuantConnect.Logging;
+using QuantConnect.Util;
+
+namespace QuantConnect.ToolBox.PsychSignalDataConverter
+{
+    public class PsychSignalDataDownloader
+    {
+        private readonly string _apiKey;
+        private readonly string _dataSource;
+
+        /// <summary>
+        /// Psychsignal imposes strict API rate limits set to about one request per ten seconds
+        /// </summary>
+        private readonly RateGate _apiRateGate = new RateGate(occurrences: 1, timeUnit: TimeSpan.FromSeconds(10));
+
+        /// <summary>
+        /// Base URL for the psychsignal API
+        /// </summary>
+        private readonly string _baseUrl = "https://api.psychsignal.com/v2";
+
+        /// <summary>
+        /// Destination we will write raw data to
+        /// </summary>
+        private readonly string _rawDataDestination;
+
+        /// <summary>
+        /// Maximum amount of retries per data hour
+        /// </summary>
+        public int MaxRetries = 5;
+
+        /// <summary>
+        /// Downlods data from psychsignal
+        /// </summary>
+        /// <param name="rawDataDestination">Directory we write raw data to</param>
+        /// <param name="apiKey">PsychSignal API key</param>
+        /// <param name="dataSource">Data source (e.g. stocktwits,twitter_withretweets)</param>
+        public PsychSignalDataDownloader(string rawDataDestination, string apiKey, string dataSource)
+        {
+            _rawDataDestination = rawDataDestination;
+
+            _dataSource = dataSource;
+            _apiKey = apiKey;
+        }
+
+        /// <summary>
+        /// Download the data from the given starting date to the ending date.
+        /// Note that if the ending date is in the same hour as the current time,
+        /// we will lower the <paramref name="endDateUtc" /> by one hour in order
+        /// to make sure that we only download complete, non-changing dataa
+        /// </summary>
+        /// <param name="startDateUtc">Starting date</param>
+        /// <param name="endDateUtc">Ending date</param>
+        public void Download(DateTime startDateUtc, DateTime endDateUtc)
+        {
+            var now = DateTime.UtcNow;
+            var nowHour = new DateTime(now.Year, now.Month, now.Day, now.Hour, 0, 0);
+
+            Directory.CreateDirectory(_rawDataDestination);
+
+            if (startDateUtc < now.AddDays(-15))
+            {
+                throw new ArgumentException("The starting date can only be at most 15 days from now");
+            }
+
+            // Makes sure we only get final, non-changing data by checking if the end date is greater than
+            // or equal to the current time and setting it to an hour before the current time if the condition is met
+            if (nowHour <= new DateTime(endDateUtc.Year, endDateUtc.Month, endDateUtc.Day, endDateUtc.Hour, 0, 0))
```

Comment:
awesome, thanks for adding this check 👍  -- style comment -- in general, I find it's best to keep the code logically matching the the comments. The comment says we're checking if the end date is greater than or equal to the current time but the code is checking if the current time is less than or equal to the end date. These are (of course) semantically the same, but does require the reader to perform the flip in their head and could lead to confusion.